module/os/windows/zfs/zfs_windows_zvol_scsi.c

```diff
@@ -1010,3 +1013,108 @@ wzvol_GeneralWkRtn(
 	wzvol_WkRtn(pWkParms);
 
 }
+
+/*
+** ZFS ZVOLDI
+** ZVOL Direct Interface
+*/
+
+VOID
+bzvol_ReadWriteTaskRtn(
+    __in PVOID           pWkParms          // Parm list 
+)
+{
+    NTSTATUS Status;
+    pMP_WorkRtnParms pWkRtnParms = (pMP_WorkRtnParms)pWkParms;
+    zfsiodesc_t* pIo = &pWkRtnParms->ioDesc;
+    int iores;
+
+    struct iovec iov;
+    iov.iov_base = (void*)pIo->Buffer;
+    iov.iov_len = pIo->Length;
+
+    zfs_uio_t uio;
+    zfs_uio_iovec_init(&uio, &iov, 1, pIo->ByteOffset, UIO_SYSSPACE, pIo->Length, 0);
+
+    /* Call ZFS to read/write data */
+
+    if (ActionRead == pWkRtnParms->Action) {
+	iores = zvol_os_read_zv(pWkRtnParms->zv, &uio, 0);
+    }
+    else {
+	iores = zvol_os_write_zv(pWkRtnParms->zv, &uio, 0); /* TODO add flag if FUA */
+    }
+
+    if (pIo->Cb) {
+	pIo->Cb(pIo, iores == 0 ? STATUS_SUCCESS : STATUS_UNSUCCESSFUL, TRUE);
+    }
+
+    ExFreePoolWithTag(pWkRtnParms, MP_TAG_GENERAL);
+}
+
+VOID bzvol_TaskQueuingWkRtn(
+    __in PVOID           pDummy,           // Not used.	
+    __in PVOID           pWkParms          // Parm list 
+)
+{
+    pMP_WorkRtnParms pWkRtnParms = (pMP_WorkRtnParms)pWkParms;
+    zfsiodesc_t* pIo = &pWkRtnParms->ioDesc;
+    PIO_WORKITEM pWI = (PIO_WORKITEM)ALIGN_UP_POINTER_BY(pWkRtnParms->pQueueWorkItem, 16);
+
+    UNREFERENCED_PARAMETER(pDummy);
+    IoUninitializeWorkItem(pWI);
+
+    if (pIo->Flags & ZFSZVOLFG_AlwaysPend) {
+	taskq_init_ent(&pWkRtnParms->ent);
+	taskq_dispatch_ent(zvol_taskq, bzvol_ReadWriteTaskRtn, pWkRtnParms, 0, &pWkRtnParms->ent);
+    }
+    else {
+	// bypass the taskq and do everything under workitem thread context.
+	bzvol_ReadWriteTaskRtn(pWkRtnParms);
+    }
+    // workitem freed inside bzvol_ReadWriteTaskRtn
+}
+
+NTSTATUS
+DiReadWriteSetup(
+    zvol_state_t* zv,
+    MpWkRtnAction action,
+    zfsiodesc_t* pIo)
+{
+    // SSV-19147: cannot use kmem_alloc with sleep if IRQL dispatch so get straight from NP pool.
```

Comment:
There is KM_PUSHPAGE for the situations you can't KM_SLEEP - uses a special pool. But yes, uses mutex in kmem - so probably isn't a great fit here.